pyro/infer/mcmc/hmc.py

```diff
@@ -309,13 +299,15 @@ def initial_trace(self):
             trace_log_prob_sum = self._compute_trace_log_prob(trace)
             if not torch_isnan(trace_log_prob_sum) and not torch_isinf(trace_log_prob_sum):
                 self._initial_trace = trace
+                self._initialize_step_size()  # this method also caches z and its potential energy
                 return trace
             trace = poutine.trace(self.model).get_trace(*self._args, **self._kwargs)
         raise ValueError("Model specification seems incorrect - cannot find a valid trace.")
 
     @initial_trace.setter
     def initial_trace(self, trace):
         self._initial_trace = trace
+        self._initialize_step_size()
```

Comment:
Sorry, my suggestion wouldn't work because `setup` is always called after we set the `intial_trace`, but there is a much simpler solution for this. Let us remove this from `initial_trace` completely, and just put it in `setup()` towards the very end. However, you will need to make one small change to `_initialize_step_size` :

```
    def _initialize_step_size(self):
        z = {name: node["value"].detach()
             for name, node in self._iter_latent_nodes(**self.initial_trace**)}  # not self._initial_trace
        ....
```

This will set the initial trace if it is not already specified, and should be much simpler overall.

Reply:
The failing test should pass once you make this change.

Reply:
Just pushed this to your branch - I hope that is okay!

Reply:
No worry, feel free to do so. :)